2022/Q4/artifacts/PoS-pseudocode/PoS-model.md

```diff
@@ -509,37 +520,54 @@ end_of_epoch()
-        update_total_deltas(validator_address, offset, diff_slashed_amount)
-        update_voting_power(validator_address, offset)
+    // Find the total amount deducted from the deltas due to unbonds that became active after the infraction epoch. This is used to ensure the deltas are appropriately slashed.
+    // Note: need to only do this once
+    var total_unbonded = 0
+    // Find the total unbonded from the slash epoch up to the current epoch first
+    // a..b notation determines an integer range: all integers between a and b inclusive
+    forall (epoch in slash.epoch+1..cur_epoch) do
+      forall ((unbond_start, unbond_amount) in validators[validator_address].unbond_records[epoch] s.t. unbond_start <= infraction_epoch && unbond_amount > 0)
+        var set_prev_slashes = {s | s in slashes[validator_address] && unbond_start <= s.epoch && s.epoch + unbonding_length + cubic_slash_window_width < infraction_epoch}
+        total_unbonded += compute_amount_after_slashing(set_prev_slashes, unbond_amount)
+
+    // For the future epochs, do the same as before but also update the deltas
+    var last_slash = 0
+    // Up to pipeline_length because there cannot be any unbond in a greater epoch (cur_epoch+pipeline_length is the upper bound)
+    forall (offset in 1..pipeline_length) do
+      forall ((unbond_start, unbond_amount) in validators[validator_address].unbond_records[cur_epoch + offset] s.t. unbond_start <= infraction_epoch && unbond_amount > 0) do
+        // We only need to apply a slash s if s.epoch < unbond.end - unbonding_length
+        // It is easy to see that s.epoch + unbonding_length < slash.epoch implies s.epoch < unbond.end - unbonding_length
+        // 1) slash.epoch = cur_epoch - unbonding_length
+        // 2) unbond.end = cur_epoch + offset + unbonding_length => cur_epoch = unbond.end - offset - unbonding_length
+        // By 1) s.epoch + unbonding_length < cur_epoch - unbonding_length
+        // By 2) s.epoch + unbonding_length < unbond.end - offset - 2*unbonding_length => s.epoch < unbond.end - offset - 3*unbonding_length, as required.
+        var set_prev_slashes = {s | s in slashes[validator_address] && unbond_start <= s.epoch && s.epoch + unbonding_length + cubic_slash_window_width < infraction_epoch}
+        total_unbonded += compute_amount_after_slashing(set_prev_slashes, unbond_amount)
+      var this_slash = (total_staked - total_unbonded) * total_rate
+      var diff_slashed_amount = last_slash - this_slash
+      last_slash = this_slash
+
+      // Do not over-slash or improperly slash stake that did not exist at the infraction epoch
+      var sum_post_bonds = sum{bonds[validator_address].deltas[start] s.t. start > infraction_epoch && start <= cur_epoch + offset>}
+      var validator_stake = read_epoched_field(validators[validator_address].total_deltas, cur_epoch + offset, 0)
+      var slashable_stake = validator_stake - sum_post_bonds
+      var change = min{-slashable_stake, diff_slashed_amount}
```

Comment:
This part I am not sure about it. I think we just need to compute how much from this_slash is left after applying any preceding slash. Could figure out if what it is in here is equivalent, but we should avoid iterating over bonds at the end of an epoch.

Reply:
I keep thinking about this. I believe that to implement this correctly we need to be able to answer the following question: given an epoch `e` and a validator `val`, how much stake is left at an epoch `e' > e` from the stake the validator had at `e`. I haven't been able to figure out how to answer this question with the state variables that we have and without iterating over bonds/unbonds.

My guess is that you code tries to do something like that, but I am not sure it succeeds on doing so. My concern is related to unbonds that remove bonds but are only effective at the `pipeline_length` offset.

Reply:
I am proposing the following: keep a new variable `total_bonded` for each validator which is a map from Epoch to int. 
We update the variable when bonding as follows:
`validators[validator_address].total_bonded[cur_epoch+pipeline_length] += amount`

Given a `offset`, we use this state variable as follows to substitute lines 564.
```go
var recent_unbonds = 0
forall (let [bond_start, amount] of validators[validator_address].unbond_records[cur_epoch+offset] s.t. amount > 0 &&
                                                                      bond_start > infraction_epoch) do
  recent_unbonds += amount                                                                    
sum_post_bonds = sum_post_bonds + (validators[validator_address].total_bonded[cur_epoch+offset] - recent_unbonds)

```
Note that `sum_post_bonds` is always `>= 0` (we are only interested in the bonds that were added after the infraction epoch and we subtract how much of those have been unbonded) and that `sum_post_bonds` is computed iteratively within the for all loops. How does this look for you? It is similar to your code but it addresses to issues:
- Does not iterate over bonds
- I am not 100% sure, but I think your code has the issue of not accounting for unbonds properly.

Reply:
Does the commit 01a5028596df73c690401f783840c902183b3bd0 satisfy everything you had in mind here?

Reply:
I think we should start computing `sum_post_bonds` in the previous `forall`: the one that goes from `slash.epoch+1` to `cur_epoch`. Otherwise, we would be possibly missing a bunch of bonds.

Reply:
fixed in 65f733515f93d9256a48a95d586b8fd7ba3c8ea5

Reply:
Have you thought of an example where the previous code fails to account for things but your proposed code successfully does? I haven't been able to do so yet.

In your code, `total_bonded` is not decremented when unbonding, but is rather effectively done so when we are computing slashes with `recent_unbonds`. On the contrary, the original code seems to do this more eagerly when unbonding rather. I'm still wondering about a difference between doing it while slashing vs. while unbonding.

Reply:
Ah, so this is not something that would occur in our current implementation. Rather, when unbonding, the change would be `total_bonded[e'] -= x`; `total_bonded` works exactly the same as `bonds[delegator_address][validator_address]`, except folds in all of the delegators such that it is `total_bonded[validator_address]`. Thus, `total_bonded` would never go negative as long as `bonds` cannot go negative.

I'll try to point out some relevant code in Namada that shows what we are doing. Hope this can possibly be helpful, though I understand you probably haven't seen any of our Rust code before.

In Namada, we are using something called [`global_bond_handle`](https://github.com/anoma/namada/blob/main/proof_of_stake/src/lib.rs#L304-L308) to hold the sum of all bonds to a validator in the same way as the bonds that are prefixed by delegator and validator address. This is the object analagous to `total_bonded`.

When we bond tokens, we update both the normal bonds and the global bonds in the same way ([see code here](https://github.com/anoma/namada/blob/main/proof_of_stake/src/lib.rs#L911-L925)).

When we unbond tokens, we decrement the bond values in place for both the normal bonds and the global bonds. As long as the normal bond has a nonzero value, the global bond is also guaranteed to have nonzero value. See the relevant code [here](https://github.com/anoma/namada/blob/main/proof_of_stake/src/lib.rs#L1612-L1659), particularly [these lines](https://github.com/anoma/namada/blob/main/proof_of_stake/src/lib.rs#L1629-L1630).

Reply:
I will put this in pseudocode right now to share.

Reply:
> Tried to put this in pseudocode quickly here: https://github.com/informalsystems/partnership-heliax/commit/ac99310bed235cbb5c8806ec44077a62bb975994

That may not work. Let's say you bond `x` tokens at epoch `e` and unbond those tokens at epoch `e+2` and you are computing `sum_post_bonds` at epoch `e`. Then the stake at `e` includes those `x` tokens but your `global_bonds[validator_address].deltas[start]` won't: you subtract them when unbonding. This looks incorrect to me.

Reply:
Sorry, I think I need more context for your example here... if I'm computing `sum_post_bonds` at epoch `e` in my version of things, then there is no way that an unbond of tokens can remove stake at `e+2` because the validator will be frozen already by epoch `e` when the unbond tx is submitted (only computing `sum_post_bonds` for future epochs relative to the current).

Reply:
- Assume that `unbonding_length=4`, pipeline_length=3 and `cubic_offset=1`
- A validator misbehaves at `e`, then it will be processed at the end of `e+unbonding_length+cubic_offset=e+5`
- The infraction is submitted at `e+unbonding_length=e+4`
- A user delegates `x` tokens to the validator at `e+2` and unbonds them at `e+unbonding_length=e+4` (before the infraction is submitted). 
- When processing the infraction at `e+5`, we would compute `sum_post_bonds` for `e+6`. By that epoch, the unbond has not materialized, but `global_bonds[e+2]` accounts for it.

I know those are not typical values for the `unbonding_length` and `pipeline_length`, but I'd rather not have anything in the design that depends on the value of those.

Reply:
Thx for this example. I agree that your proposed changes do indeed solve a problem that currently exists. I wrote up a comparison of the two approaches that demonstrates where the current one fails and where the proposed one succeeds: https://hackmd.io/yRsLbn5IT8C2xkj8csSZMg?both